xds/internal/balancer/gracefulswitch/gracefulswitch_test.go

```diff
@@ -0,0 +1,1125 @@
+	}
+	// This will be used to update the Graceful Switch Balancer. This update
+	// should simply be forwarded down to the Current Load Balancing Policy.
+	ccs := balancer.ClientConnState{
+		BalancerConfig: mockBalancerConfig{},
+	}
+
+	// Updating ClientConnState should forward the update exactly as is to the
+	// current balancer.
+	if err := gsb.UpdateClientConnState(ccs); err != nil {
+		t.Fatalf("Balancer.UpdateClientConnState(%v) failed with error: %v", ccs, err)
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := gsb.balancerCurrent.Balancer.(*mockBalancer).waitForClientConnUpdate(ctx, ccs); err != nil {
+		t.Fatalf("error in ClientConnState update: %v", err)
+	}
+}
+
+// TestTwoBalancersSameType tests the scenario where there is a graceful switch
+// load balancer setup with a current and pending Load Balancer of the same
+// type. Any ClientConn update should be forwarded to the current if no pending,
+// and the only the pending if the graceful switch balancer contains both a
+// current and pending. The pending Load Balancer should also swap into current
+// whenever it updates with a connectivity state other than CONNECTING.
+func (s) TestTwoBalancersSameType(t *testing.T) {
+	tcc, gsb := setup(t)
+	// This will be used to update the Graceful Switch Balancer. This update
+	// should simply be forwarded down to either the Current or Pending Load
+	// Balancing Policy.
+	ccs := balancer.ClientConnState{
+		BalancerConfig: mockBalancerConfig{},
+	}
+
+	gsb.SwitchTo(mockBalancerBuilder1{})
+	gsb.UpdateClientConnState(ccs)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := gsb.balancerCurrent.Balancer.(*mockBalancer).waitForClientConnUpdate(ctx, ccs); err != nil {
+		t.Fatalf("error in ClientConnState update: %v", err)
+	}
+
+	// The current balancer reporting READY should cause this state
+	// to be forwarded to the Client Conn.
+	gsb.balancerCurrent.Balancer.(*mockBalancer).updateState(balancer.State{
+		ConnectivityState: connectivity.Ready,
+		Picker:            &neverErrPicker{},
+	})
+
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for a part UpdateState call on the ClientConn")
+	case state := <-tcc.NewStateCh:
+		if state != connectivity.Ready {
+			t.Fatal("wanted connectivity state ready")
+		}
+	}
+
+	select {
+	case <-ctx.Done():
+		t.Fatalf("timeout while waiting for a part UpdateState call on the ClientConn")
+	case picker := <-tcc.NewPickerCh:
+		// Should receive a never err picker.
+		_, err := picker.Pick(balancer.PickInfo{})
+		if err != nil {
+			t.Fatalf("Graceful Switch Balancer should have sent a never error picker, which is a picker that never errors, instead received err %v", err)
```

Comment:
Someone running the test and seeing an error message should be able to figure out to a good degree why the test is failing. This error message is very hard to read and does not give a good indication of why the test has failed.

Reply:
Switched this to "ClientConn should have received a never err picker from an UpdateState call".